mirrord/agent/src/steal/connection.rs

```diff
@@ -364,6 +552,51 @@ impl TcpConnectionStealer {
         }
     }
 
+    fn handle_response_reconstruction_fail(err: hyper::http::Error) -> hyper::http::Error {
+        error!("Could not reconstruct http response: {err:?}");
+        // TODO: What should happen? Currently just skipping that response, which would
+        //       also spoils the rest of the stream. Should hopefully never happen.
+        debug_assert!(false);
+        err
+    }
+
+    /// Forward an HTTP response to a stolen HTTP request from the layer to the back to the browser.
+    ///
+    ///                         _______________agent_______________
+    /// Local App --> Layer --> ClientConnectionHandler --> Stealer --> Browser
+    ///                                                             ^- You are here.
+    #[tracing::instrument(
+        level = "trace",
+        skip(self),
+        fields(response_senders = ?self.http_response_senders.keys()),
+    )]
+    async fn http_response(&mut self, response: HttpResponse) -> Result<()> {
+        match self
+            .http_response_senders
+            .remove(&(response.connection_id, response.request_id))
+        {
+            None => {
+                warn!("Got unexpected http response. Not forwarding.");
+                Ok(())
+            }
+            Some(response_tx) => {
+                if let Err(resp) = response_tx.send(
+                    response
+                        .response
+                        .try_into()
+                        .map_err(Self::handle_response_reconstruction_fail)?,
+                ) {
+                    warn!(
```

Comment:
It's a different error, the map_err deals with the result of `response.try_into`, and this deals with the result of `send`.

Reply:
I meant, why do you use map_err for the first error but an if clause for the second error? Why don't you chain combinators for both?

Reply:
Chained everywhere https://github.com/metalbear-co/mirrord/pull/843/commits/12d32430210967453c784d6a5ffcfad563203595
Also changed this function not to return a result, so that we never stop the agent because of an error in a response.